### PR TITLE
Add sharded queries metrics

### LINKED ISSUE #5606: Add metrics to PromQLShardingMiddleware
### Is your proposal related to a problem?

Now we don't have any instrumentation to the `PromQLShardingMiddleware` so we might miss some very important stats.
I'd like to see metrics like Number of shardable queries and non-shardable queries.

### Describe the solution you'd like

The implementation should be straightforward to add two counters to the middleware

### PR DESCRIPTION
**Pull request by idoqo**
This commit adds basic metrics to track the number of shardable and non-shardable queries processed by `PromQLShardingMiddleware`.

Fixes #5606 

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Verification
Newly added metrics show up in `http://<query-frontend-address>/metrics`

### CODE REVIEW

**Comment by fpetkovski** [comment]
This is looking really good. I just wonder if we should make the result a label so that ration calculation can be easier. E.g
```
rate(frontend_sharding_middleware_queries_total{shardable="true"}[5m])
/
rate(frontend_sharding_middleware_queries_total[5m])

**Comment by yeya24** [comment]
Agree, let's make it shardable a label and keep one metric only

**Review by fpetkovski — COMMENTED** [review]
Nice work, thanks!

**Comment by yeya24** [comment]
We need another rebase.

File: CHANGELOG.md [hunk 1]
```diff
@@ -49,6 +49,7 @@ We use *breaking :warning:* to mark changes that are not backward compatible (re
 - [#5561](https://github.com/thanos-io/thanos/pull/5561) Query Frontend: Support instant query vertical sharding.
 - [#5453](https://github.com/thanos-io/thanos/pull/5453) Compact: Skip erroneous empty non `*AggrChunk` chunks during 1h downsampling of 5m resolution blocks.
 - [#5607](https://github.com/thanos-io/thanos/pull/5607) Query: Support custom lookback delta from request in query api.
+- [#5561](https://github.com/thanos-io/thanos/pull/5650) Query Frontend: Add sharded queries metrics.
```
Comment (yeya24) [inline CHANGELOG.md:52]: Please add this change to the `Added` section in `Unreleased`.

**Review by yeya24 — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (idoqo) [reply CHANGELOG.md]: right - probably messed it up during a merge, fixed it now.

File: CHANGELOG.md [hunk 2]
```diff
@@ -16,6 +16,7 @@ We use *breaking :warning:* to mark changes that are not backward compatible (re
 
 ### Added
 * [#5654](https://github.com/thanos-io/thanos/pull/5654) Query: add `--grpc-compression` flag that controls the compression used in gRPC client. With the flag it is now possible to compress the traffic between Query and StoreAPI nodes - you get lower network usage in exchange for a bit higher CPU/RAM usage.
+- [#5561](https://github.com/thanos-io/thanos/pull/5650) Query Frontend: Add sharded queries metrics.
```
Comment (yeya24) [inline CHANGELOG.md:19]: Should be `#5650` rather than `#5561`. That's the last thing then we can merge :)

**Review by yeya24 — APPROVED** [review]
LGTM

**Pull request merged**